README.md
# jsort [![Build Status](https://travis-ci.org/fknussel/jsort.svg)](https://travis-ci.org/fknussel/jsort)

Simple array sorting utility supporting multiple data types.

## Install

```
npm install --save jsort
```

## Usage

jsort has built-in support for both AMD and CommonJS module formats.

### jsort.`<dataType>(array [, key])`

Available data types include: `text`, `numeric`, `currency` and `date`. Default sorting order is `ASC`.

```js
jsort.text(['Argentina', 'Australia', 'New Zealand', 'Ireland', 'Canada']);
jsort.numeric([2.01, 1.3555, 3, 1990]);
jsort.currency(['$1,726', '$3,021,726.00', '$120.75']);
jsort.date(['2015-01-31', '01/30/2015', 'Sat, Jan 3, 2015', 'January 31, 2014']);
```

When a key gets passed in, jsort assumes the collection contains objects and will try to sort them by the key provided:

```js
jsort.numeric([{name: 'Homer', age: 40}, {name: 'Marge', age: 35}], 'age');
```

### jsort.`<dataType>(array [, key]).reverse()`

Reverse sort any array (`DESC`) by using the `.reverse()` method:

```js
jsort.text(['I', 'am', 'Yoda']).reverse(); // returns ['Yoda', 'I', 'am']
```

## Run Tests

```
grunt test
```

If you don't want to install `grunt-cli` globally, you can also run:

```
npm test
```

## Build

This Grunt task will run [JSHint](https://github.com/jshint/jshint) on all of the code, run all of the unit tests and also wrap the module into the [simplified CommonJS wrapper format](https://github.com/amdjs/amdjs-api/wiki/AMD#simplified-commonjs-wrapping-).

```
grunt build
```

## Release Versions

This repo uses [grunt-bump](https://github.com/gruntjs/grunt-bump) and Semantic Versioning to version and tag releases. To release a new version, run the appropriate command:

```
grunt release:major       # bump major version, eg. 1.0.2 -> 2.0.0
grunt release:minor       # bump minor version, eg. 0.1.3 -> 0.2.0
grunt release:patch       # bump patch version, eg. 0.0.1 -> 0.0.2
grunt release:prerelease  # bump pre-release version, eg. 1.0.0 -> 1.0.0-1
```

Make sure to push tags:

```
git push --tags origin master
```

Publish the package to [npm's public registry](https://www.npmjs.com/):

```
npm publish
```

To make sure everything worked just fine, go to [http://npmjs.com/package/jsort](http://npmjs.com/package/jsort).

**Heads up!** To publish, you must have a user on the npm registry. If you don't have one, create it with `npm adduser`. If you created one on the site, use `npm login` to store the credentials on the client. You can use `npm config ls` to ensure that the credentials are stored on your client. Check that it has been added to the registry by going to [http://npmjs.com/~](http://npmjs.com/~).

## Semantic Versioning

Given a version number `MAJOR.MINOR.PATCH`, increment the:

1. `MAJOR` version when you make incompatible API changes,
2. `MINOR` version when you add functionality in a backwards-compatible manner, and
3. `PATCH` version when you make backwards-compatible bug fixes.

Additional labels for pre-release and build metadata are available as extensions to the `MAJOR.MINOR.PATCH` format.

See the [Semantic Versioning](http://semver.org/) specification for more information.

## Release History

See the [CHANGELOG](CHANGELOG.md).


CHANGELOG.md

# 1.0.2

- Merge pull request #1 from fknussel/develop `acf1d7f`
- Fix parse-date unit tests `9f5922e`
- Integrate with Travis CI `e08891c`

# 1.0.1

- Support sorting dates in a variety of formats `44a51eb`

# 1.0.0

- Update docs and code samples with new features `ef76b2a`
- Add support for any currency `8726818`
- Add ability to sort collections of objects by key `03b7903`
- Add jshint and build Grunt tasks `cc55b97`
- Fix Jest configuration `695538b`
- Update module API `169ebea`

# 0.1.1

- Add extra unit tests `64c2510`
- Make text comparison case insensitive `2e72aad`
- Add Grunt task to wrap CommonJS code into the AMD module format `ae71abf`
- Add more examples to demo.js `3144635`

# 0.1.0

- Add code samples `6dab377`
- Add Grunt release tasks `6c6f6db`


src/demo.js
var chalk = require('chalk');
var jsort = require('./jsort');

// Strings
var countries = ['Australia', 'Argentina', 'New Zealand', 'Austria', 'Canada'];
console.log('\n' + chalk.white.bgCyan.bold('Sorting a collection of strings'));
console.log('Sorted countries: ', jsort.text(countries));
console.log('Reverse-sorted countries: ', jsort.text(countries).reverse());

// Integers
var integers = [1, 10, 11, 110, 111, 0];
console.log('\n' + chalk.white.bgCyan.bold('Sorting a collection of integers'));
console.log('Sorted integers: ', jsort.numeric(integers));
console.log('Reverse-sorted integers: ', jsort.numeric(integers).reverse());

// Floats
var floats = [1.1, 1.15, 1.20, 10.1, 11.1, 0.01];
console.log('\n' + chalk.white.bgCyan.bold('Sorting a collection of floats'));
console.log('Sorted floats: ', jsort.numeric(floats));
console.log('Reverse-sorted floats: ', jsort.numeric(floats).reverse());

// YYYY-MM-DD formatted dates
var dates = ['2015-02-04', '2015-02-02', '2017-02-01'];
console.log('\n' + chalk.white.bgCyan.bold('Sorting a collection of dates in YYYY-MM-DD format'));
console.log('Sorted dates: ', jsort.date(dates));
console.log('Reverse-sorted dates: ', jsort.date(dates).reverse());

// MM/DD/YYYY formatted dates
var moreDates = ['01/31/2015', '01/30/2015', '01/31/2016'];
console.log('\n' + chalk.white.bgCyan.bold('Sorting a collection of dates in MM/DD/YYYY format'));
console.log('Sorted dates: ', jsort.date(moreDates));
console.log('Reverse-sorted dates: ', jsort.date(moreDates).reverse());

// Currency
var sales = ['$1,726.75', '$1,726.50', '$1,720.99'];
console.log('\n' + chalk.white.bgCyan.bold('Sorting a collection of currencies'));
console.log('Sorted currencies: ', jsort.currency(sales));
console.log('Reverse-sorted currencies: ', jsort.currency(sales).reverse());

// Objects
var employees = [
  { age: 25, dob: '01/01/2010', salary: '$100,000.00', name: 'Ted' },
  { age: 10, dob: '01/01/2001', salary: '$80,000.00', name: 'Mario' },
  { age: 50, dob: '01/01/2009', salary: '$75,000.00', name: 'Homer' }
];
console.log('\n' + chalk.white.bgCyan.bold('Sorting a collection of objects by key'));
console.log('Employees sorted by name: ', jsort.text(employees, 'name'));
console.log('Employees sorted by age: ', jsort.numeric(employees, 'age'));
console.log('Employees sorted by salary: ', jsort.currency(employees, 'salary'));
console.log('Employees sorted by date of birth: ', jsort.date(employees, 'dob'));
